### PR TITLE
docs: Improved the installation instructions of the Pleroma service

### PR DESCRIPTION
**Pull request by massimo-zaniboni**
###### Motivation for this change

I installed the ``services.pleroma`` and so I acquired experience for improving the documentation.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/pleroma.xml
Comment (happysalada) [inline nixos/modules/services/networking/pleroma.xml:27]: perhaps default to 13 here ?

File: nixos/modules/services/networking/pleroma.xml
Comment (happysalada) [inline nixos/modules/services/networking/pleroma.xml:56]: perhaps add these secrets as well to the secrets file ?

**Comment by happysalada** [comment]
Thank you for this change! Documentation is critical and we need to improve it?

Could you not include the master merge commit ?

@petabyteboy @NinjaTrappeur 
Perhaps you have some additional thoughts ?

File: nixos/modules/services/networking/pleroma.xml
Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.xml:56]: Agree.

File: nixos/modules/services/networking/pleroma.xml
Comment (NinjaTrappeur) [inline nixos/modules/services/networking/pleroma.xml:73]: Same. This should live in the path pointed by `secretConfigFile`, not in the store.

You can have a loot at the config of the NixOS test to see an example https://github.com/NixOS/nixpkgs/blob/3ab8ce12c2db31268f579c11727d9c63cfee2eee/nixos/tests/pleroma.nix#L123

File: nixos/modules/services/networking/pleroma.xml
Comment (NinjaTrappeur) [inline nixos/modules/services/networking/pleroma.xml:152]: Meh. Pleroma only requires a single vhost to operate. I don't see why you'd require a wildcard certificate just for this service.

Let's encourage users to use good security practices and only issue the certificate for the domain name we actually need.

**Review by picnoir — COMMENTED** [review]
Thanks a lot for this, this is a clear improvement!

Overall LGTM after we adress some comments I left in the review.

**Comment by NinjaTrappeur** [comment]
CCing @yu-re-ka as they're also a pleroma maintainer.

**Comment by happysalada** [comment]
Thank you for finding her, the maintainers list has yuka and I wasn't sure it was her.

File: nixos/modules/services/networking/pleroma.xml
Comment (massimo-zaniboni) [reply nixos/modules/services/networking/pleroma.xml]: done

File: nixos/modules/services/networking/pleroma.xml
Comment (massimo-zaniboni) [reply nixos/modules/services/networking/pleroma.xml]: To be fair these were not "secrets" but only place-holder strings, overriden by ``/var/lib/pleroma/secrets.exs`` real secrets. But you are right in any case, because it was not an optimal way to split secrets, so now I left secrets only on ``/var/lib/pleroma/secrets.exs``, and I used also a "fake-secret" for avoiding GitHub complaints about potentially exposed secrets. So I improved this part.

File: nixos/modules/services/networking/pleroma.xml
Comment (massimo-zaniboni) [reply nixos/modules/services/networking/pleroma.xml]: done

File: nixos/modules/services/networking/pleroma.xml
Comment (massimo-zaniboni) [reply nixos/modules/services/networking/pleroma.xml]: Ops! I miss interpreted the section of NixOS manual about ACME/Let's Encrypt https://nixos.org/manual/nixos/stable/index.html#module-security-acme-configuring . I thought this setting was mandatory. Instead it suffices a lot less!!!

I fixed this on my server, and tested creating new domains and it works. So I committed a simpler and safer version.

**Comment by massimo-zaniboni** [comment]
> Thank you for this change! Documentation is critical and we need to improve it?
> Could you not include the master merge commit ?

I have no idea of how to avoid this now. Probably when someone merge, he can merge a squash version of my commits.

**Comment by happysalada** [comment]
you can ammend commits in a PR with `git rebase -i --autosquash HEAD~2`
the i flag is for interactive and the autosquash is not mandatory.
When you have changed your branch history you can `git push --force`.
let me know if any of those doesn't make sense.
if any of those is a pain, I don't mind merging a squashed version, no problems on my side.

File: nixos/modules/services/networking/pleroma.xml
Comment (happysalada) [inline nixos/modules/services/networking/pleroma.xml:130]: if you are interested, instead of documenting those, a PR could be done about changing the nixos module and add a configuration `services.plerorma.nging.enable = true;` that would enable all of those settings.
this is exactly what is done here https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/services/web-apps/discourse.nix#L772
I understand that this is a little more involved, so no worries if you want to address this in a separate PR or not address this at all (both are completely fine with me).

Comment (massimo-zaniboni) [reply nixos/modules/services/networking/pleroma.xml:130]: I have no time to code and test the ``services.pleroma.nginx.enable = true;`` but inspired from your link and from the configuration I'm using in production, I improved the Nginx configuration. In particular ``etags`` is very important for caching images and similar content, and it is probably mandatory on a Pleroma server.

**Comment by massimo-zaniboni** [comment]
> if any of those is a pain, I don't mind merging a squashed version, no problems on my side.

I prefer a squashed history on your side (i.e. NixOS repo), but a complete history of intermediate commits on the PR side (i.e. my temporary fork/branch). So, if it is fast for you, you can squash and merge the PR when you consider it good enough. I will mantain my history.

**Comment by happysalada** [comment]
@massimo-zaniboni squashing here is no problem.

Can I request a final check from @NinjaTrappeur when you have a moment ?

File: nixos/modules/services/networking/pleroma.xml
Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.xml]: I get it, but a reader might assume `SECRET` is a placeholder they need to fill with the actual values.

File: nixos/modules/services/networking/pleroma.xml
Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.xml:130]: I personally dislike this kind of automagic module webserver configuration. From my experience, it tends to backfire quite quickly.

**Comment by NinjaTrappeur** [comment]
I'm busy ATM. I'll have a further look later in the week.

**Comment by happysalada** [comment]
No rush, I'll wait for your approval before merging. Thank you for spending time on this!

**Review by picnoir — APPROVED** [review]
Massive thanks for your this doc overhaul @massimo-zaniboni, it was definitely needed!

I'm also very sorry about the latency here. It fell out of my brain at some point, may bad, my fault.

The Nginx conf turned out being alright.

I hope this won't frustrate you too much and turn you out of the NixOS project. I **really** hope to see you contribute to this project again in the future :)

I'm going to squash your commits a bit before merging them.

**Comment by NinjaTrappeur** [comment]
Arf, Github is preventing me to push to your master branch (the one targeted by this PR).

I'm going to push your commit on a branch I can write on in a bit.

**Comment by NinjaTrappeur** [comment]
I rebased everything for you there: https://github.com/NixOS/nixpkgs/pull/138281.

**Comment by massimo-zaniboni** [comment]
> I hope this won't frustrate you too much and turn you out of the NixOS project. I really hope to see you contribute to this project again in the future :)

Don't worry: I have a lot of things queued myself, so I understand your situation!!! :-)

**Comment by NinjaTrappeur** [comment]
Merged via a3d011eebcee9f70463a68c28b2252073336ed54

**Comment by NinjaTrappeur** [comment]
And now it's merged, I realize I botched the commit message by forgetting to remove squashed messages :disappointed:

**Pull request closed without merging**
